## Group Discussion
### Using transcriptomics to study developmental in non-model systems

Transcriptomics has opened up the possibility of studying development and developmental plasticity in non-model organisms. For this discussion, we will read some recent papers and discuss how -omics have been used to answer biological questions.

---
## Papers

* Casasa, S., Zattara, E.E. & Moczek, A.P. (2020). Nutrition-responsive gene expression and the developmental evolution of insect polyphenism. Nature Ecology & Evolution, 4, 1–12. [PDF](./Casasa.pdf){:target="_blank"}

* Liedtke, H.C., Harney, E. & Gomez-Mestre, I. (2021). Cross-species transcriptomics uncovers genes underlying genetic accommodation of developmental plasticity in spadefoot toads. Molecular Ecology, 30, 2220–2234. [PDF](./Liedtke.pdf){:target="_blank"}

* Stuckert, A.M.M., Moore, E., Coyle, K.P., Davison, I., MacManes, M.D., Roberts, R. & Summers, K. (2019). Variation in pigmentation gene expression is associated with distinct aposematic color morphs in the poison frog Dendrobates auratus. BMC Evolutionary Biology, 19, 1–15.[PDF](./Stuckert.pdf){:target="_blank"}

* Zhu, W., Liu, L., Wang, X., Gao, X., Jiang, J. & Bin Wang. (2019). Transcriptomics reveals the molecular processes of light-induced rapid darkening of the non-obligate cave dweller Oreolalax rhodostigmatus (Megophryidae, Anura) and their genetic basis of pigmentation strategy. BMC Genomics, 19, 1–13. [PDF](./Zhu.pdf){:target="_blank"}

* Gessler, T. B., Wu, Z., & Valenzuela, N. (2023). Transcriptomic thermal plasticity underlying gonadal development in a turtle with ZZ/ZW sex chromosomes despite canalized genotypic sex determination. Ecology and Evolution, 13(2), e9854. [PDF](./Gessler.pdf){:target="_blank"}

---

## Focus on:

*	What is the biological question/aim of the paper?
* How is transcriptomics (or other -omics) useful for answering this aims?
* What is the experimental design:
  * What tissues?
  * What treatments?
  * How many biological replicates?
  * Is there a "control" vs. "treatment", or "phenotype 1" vs "phenotype 2"?
* Is there a reference genome available?
* What software/methods are used? For:
  * Transcriptome assembly
  * Reads quantification
  * Differential gene expression
  * Functional enrichment
* How are genes of interest annotated?
